designs/2024-per-rule-autofix-configuration/README.md

```diff
@@ -0,0 +1,406 @@
+probably located before the [Disabling rules](https://eslint.org/docs/latest/use/configure/rules#disabling-rules) 
+section, so that it can be directly referenced.
+
+### `linterOptions.reportUnusedDisableDirectives`
+
+A note should be added in the
+[Configuration Files > Reporting Unused Disable Directives](https://eslint.org/docs/latest/use/configure/configuration-files#reporting-unused-disable-directives)
+section that autofix behavior can be configured using the extended format.
+
+## Drawbacks
+
+<!--
+    Why should we *not* do this? Consider why adding this into ESLint
+    might not benefit the project or the community. Attempt to think 
+    about any opposing viewpoints that reviewers might bring up. 
+
+    Any change has potential downsides, including increased maintenance
+    burden, incompatibility with other tools, breaking existing user
+    experience, etc. Try to identify as many potential problems with
+    implementing this RFC as possible.
+-->
+
+The primary drawback is the complexity of normalizing configuration formats 
+across the codebase, which requires careful implementation across multiple packages (?).
+
+It's important to ensure nothing breaks during this transition.
+
+## Backwards Compatibility Analysis
+
+<!--
+    How does this change affect existing ESLint users? Will any behavior
+    change for them? If so, how are you going to minimize the disruption
+    to existing users?
+-->
+
+This feature is fully backwards compatible:
+
+* `autofix` defaults to `true`, matching current behavior; 
+* Internal normalization does not affect user configurations or existing workflows.
+
+## Alternatives
+
+<!--
+    What other designs did you consider? Why did you decide against those?
+
+    This section should also include prior art, such as whether similar
+    projects have already implemented a similar feature.
+-->
+
+### Use of a 3rd-party plugin
+
+<https://www.npmjs.com/package/eslint-plugin-no-autofix> is a tool that exists to currently work around this limitation of ESLint, but it is not perfect.
+
+1. It is an extra third-party dependency, with its own potential maintenance issues (having to keep up with ESLint, separate dependencies that can fall out of date, obsolete, unsecure, etc.)
+2. It may not work in all environments. For example, pre-commit.ci: <https://github.com/aladdin-add/eslint-plugin/issues/98>
+3. It may not work correctly with all third-party rules: <https://github.com/eslint-community/eslint-plugin-eslint-comments/issues/234>
+
+## Open Questions
+
+<!--
+    This section is optional, but is suggested for a first draft.
+
+    What parts of this proposal are you unclear about? What do you
+    need to know before you can finalize this RFC?
+
+    List the questions that you'd like reviewers to focus on. When
+    you've received the answers and updated the design to reflect them, 
+    you can remove this section.
+-->
+
+- **Should there be legacy format support?** \
```

Comment:
I'm unclear what this question means. What is "legacy format"?

Reply:
`.eslintrc`-style configuration format should not support the new rule configuration format.

> While looking through the code, I noticed that the `eslint` package itself relies on some functions from the [@eslint/eslintrc](https://github.com/eslint/eslintrc) package ([example](https://github.com/eslint/eslint/blob/0f49329b4a7f91714f2cd1e9ce532d32202c47f4/lib/eslint/eslint.js#L20)), and those would need to be updated for the new format to work properly.

We should update the `eslint` package to not use functions from the `@eslint/eslintrc` package in code that handles the new flat config format.

Reply:
> We should update the `eslint` package to not use functions from the `@eslint/eslintrc` package in code that handles the new flat config format.

I think I got most of this with the `Config` refactor. At the very least, I believe the relevant methods now exist on `Config`.

> While looking through the code, I noticed that the `eslint` package itself relies on some functions from the [@eslint/eslintrc](https://github.com/eslint/eslintrc) package ([example](https://github.com/eslint/eslint/blob/0f49329b4a7f91714f2cd1e9ce532d32202c47f4/lib/eslint/eslint.js#L20)), and those would need to be updated for the new format to work properly.

I'll take a look at this.